client_test.go

```diff
@@ -5167,40 +5179,150 @@ func TestInsertParamsFromJobArgsAndOptions(t *testing.T) {
+			ExcludeKind: true,
+		}
+
+		expectedKey := dbunique.UniqueKey(archetype.Time, internalUniqueOpts, params)
+
+		require.Equal(t, expectedKey, params.UniqueKey)
+		require.Equal(t, internalUniqueOpts.StateBitmask(), params.UniqueStates)
+	})
+
+	t.Run("UniqueOptsCustomStates", func(t *testing.T) {
+		t.Parallel()
+
+		archetype := riversharedtest.BaseServiceArchetype(t)
+		archetype.Time.StubNowUTC(time.Now().UTC())
+
+		states := []rivertype.JobState{
+			rivertype.JobStateAvailable,
+			rivertype.JobStatePending,
+			rivertype.JobStateRetryable,
+			rivertype.JobStateRunning,
+			rivertype.JobStateScheduled,
+		}
+
+		uniqueOpts := UniqueOpts{
+			ByPeriod: 10 * time.Second,
+			ByQueue:  true,
+			ByState:  states,
+		}
+
+		params, resultUniqueOpts, err := insertParamsFromConfigArgsAndOptions(archetype, config, noOpArgs{}, &InsertOpts{UniqueOpts: uniqueOpts}, true)
+		require.NoError(t, err)
+		require.Nil(t, resultUniqueOpts)
+		internalUniqueOpts := &dbunique.UniqueOpts{
+			ByPeriod: 10 * time.Second,
+			ByQueue:  true,
+			ByState:  states,
+		}
+
+		expectedKey := dbunique.UniqueKey(archetype.Time, internalUniqueOpts, params)
+
+		require.Equal(t, expectedKey, params.UniqueKey)
+		require.Equal(t, internalUniqueOpts.StateBitmask(), params.UniqueStates)
+	})
+
+	t.Run("UniqueOptsV1", func(t *testing.T) {
 		t.Parallel()
 
+		archetype := riversharedtest.BaseServiceArchetype(t)
+		archetype.Time.StubNowUTC(time.Now().UTC())
+
 		uniqueOpts := UniqueOpts{
+			ByArgs:   true,
+			ByPeriod: 10 * time.Second,
+			ByQueue:  true,
+			// This list of custom states (without pending, scheduled, running, etc.) is only valid for v1 unique opts:
+			ByState: []rivertype.JobState{rivertype.JobStateAvailable, rivertype.JobStateCompleted},
+		}
+
+		params, resultUniqueOpts, err := insertParamsFromConfigArgsAndOptions(archetype, config, noOpArgs{}, &InsertOpts{UniqueOpts: uniqueOpts}, false)
+		require.NoError(t, err)
+		require.NotNil(t, resultUniqueOpts)
+		internalUniqueOpts := &dbunique.UniqueOpts{
 			ByArgs:   true,
 			ByPeriod: 10 * time.Second,
 			ByQueue:  true,
 			ByState:  []rivertype.JobState{rivertype.JobStateAvailable, rivertype.JobStateCompleted},
 		}
+		require.Equal(t, internalUniqueOpts, resultUniqueOpts)
+
+		require.Nil(t, params.UniqueKey)
+		require.Zero(t, params.UniqueStates)
+
+		// In a bulk insert, this should be explicitly blocked:
+		params, resultUniqueOpts, err = insertParamsFromConfigArgsAndOptions(archetype, config, noOpArgs{}, &InsertOpts{UniqueOpts: uniqueOpts}, true)
+		require.ErrorContains(t, err, "bulk inserts do not support advisory lock uniqueness and cannot remove required states")
+		require.Nil(t, params)
+		require.Nil(t, resultUniqueOpts)
+	})
+
+	t.Run("UniqueOptsWithPartialArgs", func(t *testing.T) {
+		t.Parallel()
+
+		uniqueOpts := UniqueOpts{ByArgs: true}
+
+		type PartialArgs struct {
+			JobArgsStaticKind
+			Included bool `json:"included" unique:"true"`
+			Excluded bool `json:"excluded"`
```

Comment:
I had another design idea on this. Instead of `unique:"true"`, what if this was something like `river:"unique"`? There would be room to apply additional comma-separated options in that list, such as for batching or sequence keys.

I think I like that more than having a bunch of separate tags with vague names like `unique`, `batch`, etc. Those could be prefixed with `river_` of course but that'd be long and redundant, plus not a particularly efficient way to declare boolean attrs.

Reply:
Changed to `river:"unique"` so better & more cleanly support future options. Seems less vague/generic too which is nice.